src/graph.rs

```diff
@@ -130,6 +167,843 @@ impl Handle<Symbol> {
     }
 }
 
+//-------------------------------------------------------------------------------------------------
+// Files
+
+/// A source file that we have extracted stack graph data from.
+///
+/// It's up to you to choose what names to use for your files, but they must be unique within a
+/// stack graph.  If you are analyzing files from the local filesystem, the file's path is a good
+/// choice.  If your files belong to packages or repositories, they should include the package or
+/// repository IDs to make sure that files in different packages or repositories don't clash with
+/// each other.
+pub struct File {
+    /// The name of this source file.
+    pub name: String,
+}
+
+impl StackGraph {
+    /// Adds a file to the stack graph, ensuring that there's only ever one file with a particular
+    /// name in the graph.
+    pub fn add_file<S: AsRef<str> + ?Sized>(&mut self, name: &S) -> Handle<File> {
+        let name = name.as_ref();
+        if let Some(handle) = self.file_handles.get(name) {
+            return *handle;
+        }
+        let name_value = name.to_string();
+        let file = File {
+            name: name_value.clone(),
+        };
+        let handle = self.files.add(file);
+        self.file_handles.insert(name_value, handle);
+        handle
+    }
+}
```

Comment:
I favor the return type `Handle<File>` because of its streamlined and simple shape, but it makes me wonder if there's value in encoding the invariant that files are unique to a StackGraph in the return type? Something like e.g. `Either<FileExistsError, Handle<File>>` might help communicate this invariant for users and possibly avoid surprises.

I'm not sure if that's worth the complexity this kind of result type introduces (i.e. extra layer of unwrapping, `match` to guard against the error condition, etc.), but it would prevent potentially mishandling the `add_file` operation. Curious what you think?

Reply:
That's a good suggestion!  I made `add_file` return a `Result<Handle<File>, Handle<File>>`, so that you can tell whether the file already existed, and get its handle either way.  And I added a new `get_or_create_file` method for the case where you don't care whether it already exists, and just want the handle.